flare/mgp/mapxb.py

```diff
@@ -0,0 +1,655 @@
+import logging, warnings
+import numpy as np
+import multiprocessing as mp
+
+from copy import deepcopy
+from math import ceil, floor
+from scipy.linalg import solve_triangular
+from typing import List
+
+from flare.env import AtomicEnvironment
+from flare.kernels.utils import from_mask_to_args, str_to_kernel_set
+from flare.gp import GaussianProcess
+from flare.gp_algebra import partition_vector, energy_force_vector_unit, \
+    force_energy_vector_unit, energy_energy_vector_unit, force_force_vector_unit,\
+    _global_training_data, _global_training_structures
+from flare.parameters import Parameters
+from flare.struc import Structure
+
+from flare.mgp.utils import get_kernel_term, str_to_mapped_kernel
+from flare.mgp.splines_methods import PCASplines, CubicSpline
+
+global_use_grid_kern = True
+
+class MapXbody:
+    def __init__(self,
+                 grid_num: List,
+                 lower_bound: List or str='auto',
+                 upper_bound: List or str='auto',
+                 svd_rank = 'auto',
+                 species_list: list=[],
+                 map_force: bool=False,
+                 GP: GaussianProcess=None,
+                 mean_only: bool=True,
+                 container_only: bool=True,
+                 lmp_file_name: str='lmp.mgp',
+                 load_grid: str=None,
+                 lower_bound_relax: float=0.1,
+                 n_cpus: int=None,
+                 n_sample: int=100):
+
+        # load all arguments as attributes
+        self.grid_num = np.array(grid_num)
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+        self.svd_rank = svd_rank
+        self.species_list = species_list
+        self.map_force = map_force
+        self.mean_only = mean_only
+        self.lmp_file_name = lmp_file_name
+        self.load_grid = load_grid
+        self.lower_bound_relax = lower_bound_relax
+        self.n_cpus = n_cpus
+        self.n_sample = n_sample
+        self.spc = []
+        self.spc_set = []
+
+        self.build_bond_struc(species_list)
+
+
+        bounds = [self.lower_bound, self.upper_bound]
+        self.build_map_container(bounds)
+
+        if (not container_only) and (GP is not None) and \
+                (len(GP.training_data) > 0):
+            self.build_map(GP)
+
+    def build_bond_struc(self, species_list):
+        raise NotImplementedError("need to be implemented in child class")
+
+    def get_arrays(self, atom_env):
+        raise NotImplementedError("need to be implemented in child class")
+
+    def build_map_container(self, bounds):
+        '''
+        construct an empty spline container without coefficients.
+        '''
+
+        self.maps = []
+        for spc in self.spc:
+            m = self.singlexbody((self.grid_num, bounds, spc,
```

Comment:
@YuuuuXie  maybe we should explicitly write out which variable is used for which argument. For example (grid_num=self.grid_num, bounds=bounds, ... ). So it's easier to debug if we change the init arguments of singlexbody in the future.